File: CodiceFiscale/CodiceFiscale.java

public class CodiceFiscale {
	public static void main(String[] args) {
		
	CodiceFiscale CodiceFiscale = new CodiceFiscale();
		
	System.out.println(CodiceFiscale.calcolaCodiceFiscale(
			"Mario", "Rossi", 12, 6, 1976, "M", "Bologna").equals("RSSMRA76H12A944I"));
	
	System.out.println(CodiceFiscale.calcolaCodiceFiscale(
			"Mario", "Rossi", 1, 1, 1990, "M", "Milano").equals("RSSMRA90A01F205Z"));
	
	System.out.println(CodiceFiscale.verificaCodiceFiscale(
			"Mario", "Rossi", 12, 6, 1976, "M", "Bologna", "RSSMRA76H12A94QF"));
	
	System.out.println(CodiceFiscale.verificaCodiceFiscale(
			"Mario", "Rossi", 12, 6, 1976, "M", "Bologna", "RSSMRA76H12A9Q4U"));
	
	System.out.println(CodiceFiscale.verificaCodiceFiscale(
			"Mario", "Rossi", 1, 1, 1990, "M", "Milano", "RSSMRA90A01F20RU"));
	
	System.out.println(CodiceFiscale.verificaCodiceFiscale(
			"Mario", "Rossi", 1, 1, 1990, "M", "Milano", "RSSMRA90A01F2L5K"));
	
	System.out.println(CodiceFiscale.verificaCodiceFiscale(
			"Mario", "Rossi", 1, 1, 1990, "M", "Milano", "RSSMRA90A01FN05O"));
	
	System.out.println(CodiceFiscale.verificaCodiceFiscale(
			"Mario", "Rossi", 1, 1, 1990, "M", "Milano", "RSSMRA90A0MF205R"));
	
	System.out.println(CodiceFiscale.verificaCodiceFiscale(
			"Mario", "Rossi", 1, 1, 1990, "M", "Milano", "RSSMRA90AL1F205K"));
	
	}
	
	final private String vocali = "AEIOU";
	final private String mesiTab = "ABCDEHLMPRST" ;
	final private int woman_const = 40;
	final private String milano_cod = "F205";
	final private String bologna_cod = "A944";
	// zero voglia di mettere tutta la lista comuni
	
	final private String alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
	final private String dispAlphabet = "BAKPLCQDREVOSFTGUHMINJWZYX";
	
	final private String omocodia = "LMNPQRSTUV";
	
	
	public String calcolaCodiceFiscale(String nome, String cognome, int giorno, int mese, int anno, String sesso, String comune ) {
		String partialResult = this.calcolaCognome(cognome) + this.calcolaNome(nome) + this.calcolaAnno(anno) + this.calcolaMese(mese)
			+ this.calcolaGiornoSesso(giorno, sesso) + this.calcolaComune(comune);
		
		return partialResult + this.calcolaCarControllo(partialResult);
	}
	
	public boolean verificaCodiceFiscale(String cognome, String nome, int giorno, int mese, int anno, String comune, String sesso, String CF) {
		String CalculatedCF = this.calcolaCodiceFiscale(cognome, nome, giorno, mese, anno, comune, sesso);
		
		String charCF = new String();
		String charCalculatedCF = new String();
		
		for( int i = 0; i < CF.length() -1; i++) {
			charCF = String.valueOf(CF.charAt(i));
			charCalculatedCF =  String.valueOf(CalculatedCF.charAt(i));
			
			if( !charCF.equals(charCalculatedCF)) {
				if(!charCF.equals(String.valueOf(omocodia.charAt(Integer.valueOf(charCalculatedCF))))) {
					return false;
				}
			}
		}
		
		return true;
	}
	
	
	private String calcolaComune(String comune) {
		String uppedComune = comune.toUpperCase();
		if(uppedComune.equals("BOLOGNA")) {
			return bologna_cod;
		} else if(uppedComune.equals("MILANO")) {
			return milano_cod;
		} else {
			return "FRAH";
		}
	}
	
	private String calcolaGiornoSesso(int giorno, String sesso) {
		String result = new String();
		
		if(sesso.equals("M")) {
			result = String.valueOf(giorno);
		} else {
			result = String.valueOf(giorno + woman_const);
		}
		
		if(result.length() == 1) {
			result = "0" + result;
		}
		
		return result;
	}
	
	private String calcolaMese(int mese) {
		return String.valueOf(mesiTab.charAt(mese - 1));
	}
	
	private String calcolaAnno(int anno) {
		String nascita = String.valueOf(anno % 100);
		if(nascita.length() == 1) {
			nascita = "0" + nascita;
		}
		return nascita;
	}
	
	private String calcolaCarControllo(String CFincompleto) {
		int result = 0;
		String charCF = new String();
		String ParsedCF = new String();
		
		for (int i = 0; i< CFincompleto.length() ; i++) {
			charCF = String.valueOf(CFincompleto.charAt(i));
			if(alphabet.contains(charCF)) {
				ParsedCF += charCF;
			} else {
				ParsedCF += alphabet.charAt(Integer.valueOf(charCF));
			}
		}
		
		for (int i = 0; i < ParsedCF.length(); i++) {
			charCF = String.valueOf(ParsedCF.charAt(i));
			
			if(i % 2 != 0) {
				result += alphabet.indexOf(charCF);
			} else {
				result += dispAlphabet.indexOf(charCF);
			}
			
		}
		return String.valueOf(alphabet.charAt(result % 26));
	}
	
	private String calcolaCognome(String cognome) {
		
		String uppedCognome = cognome.toUpperCase();
		String result = new String(); 
		
		if(uppedCognome.length() < 3) {
			result = uppedCognome;
			for(int i = uppedCognome.length(); i < 3; i++) {
				result += "X";
			}
			return result;
		}
		
		
		for(int i = 0; i < uppedCognome.length() && result.length() < 3; i++) {
			if(isConsonante(uppedCognome.charAt(i))) {
				result += uppedCognome.charAt(i);
			}
		}
		
		if(result.length() < 3) {
			for(int i = 0; i < uppedCognome.length() && result.length() < 3; i++) {
				if(isVocale(uppedCognome.charAt(i))) {
					result += uppedCognome.charAt(i);
				}
			}
		}
		
		return result;
	}
	
	private String calcolaNome(String nome) {
		String uppedNome = nome.toUpperCase();
		String result = new String();
		
		if(uppedNome.length() < 3) {
			result = uppedNome;
			for(int i = uppedNome.length(); i < 3; i++) {
				result += "X";
			}
			return result;
		}
		
		for(int i = 0; i < uppedNome.length() && result.length() < 4; i++) {
			if(isConsonante(uppedNome.charAt(i))) {
				result += uppedNome.charAt(i);
			}
		}
		
		if( result.length() < 4) {
			return this.calcolaCognome(uppedNome);
		}
		
		return result;
		
	}
	
	private boolean isVocale(char c) {
		return this.vocali.contains(String.valueOf(c));
	}
	
	private boolean isConsonante(char c) {
		return !this.isVocale(c);
	}
	
}

File: CodiceFiscale/CodiceFiscale.kt
import kotlin.jvm.JvmStatic

class CodiceFiscale {
    private val vocali = "AEIOU"
    private val mesiTab = "ABCDEHLMPRST"
    private val woman_const = 40
    private val milano_cod = "F205"
    private val bologna_cod = "A944"

    // zero voglia di mettere tutta la lista comuni
    private val alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    private val dispAlphabet = "BAKPLCQDREVOSFTGUHMINJWZYX"
    private val omocodia = "LMNPQRSTUV"
    fun calcolaCodiceFiscale(nome: String, cognome: String, giorno: Int, mese: Int, anno: Int, sesso: String, comune: String): String {
        val partialResult = (calcolaCognome(cognome) + calcolaNome(nome) + calcolaAnno(anno) + calcolaMese(mese)
                + calcolaGiornoSesso(giorno, sesso) + calcolaComune(comune))
        return partialResult + calcolaCarControllo(partialResult)
    }

    fun verificaCodiceFiscale(cognome: String, nome: String, giorno: Int, mese: Int, anno: Int, comune: String, sesso: String, CF: String): Boolean {
        val CalculatedCF = calcolaCodiceFiscale(cognome, nome, giorno, mese, anno, comune, sesso)
        var charCF = String()
        var charCalculatedCF = String()
        for (i in 0 until CF.length - 1) {
            charCF = CF[i].toString()
            charCalculatedCF = CalculatedCF[i].toString()
            if (charCF != charCalculatedCF) {
                if (charCF != omocodia[Integer.valueOf(charCalculatedCF)].toString()) {
                    return false
                }
            }
        }
        return true
    }

    private fun calcolaComune(comune: String): String {
        val uppedComune = comune.toUpperCase()
        return if (uppedComune == "BOLOGNA") {
            bologna_cod
        } else if (uppedComune == "MILANO") {
            milano_cod
        } else {
            "FRAH"
        }
    }

    private fun calcolaGiornoSesso(giorno: Int, sesso: String): String {
        var result = String()
        result = if (sesso == "M") {
            giorno.toString()
        } else {
            (giorno + woman_const).toString()
        }
        if (result.length == 1) {
            result = "0$result"
        }
        return result
    }

    private fun calcolaMese(mese: Int): String {
        return mesiTab[mese - 1].toString()
    }

    private fun calcolaAnno(anno: Int): String {
        var nascita = (anno % 100).toString()
        if (nascita.length == 1) {
            nascita = "0$nascita"
        }
        return nascita
    }

    private fun calcolaCarControllo(CFincompleto: String): String {
        var result = 0
        var charCF = String()
        var ParsedCF = String()
        for (i in 0 until CFincompleto.length) {
            charCF = CFincompleto[i].toString()
            if (alphabet.contains(charCF)) {
                ParsedCF += charCF
            } else {
                ParsedCF += alphabet[Integer.valueOf(charCF)]
            }
        }
        for (i in 0 until ParsedCF.length) {
            charCF = ParsedCF[i].toString()
            result += if (i % 2 != 0) {
                alphabet.indexOf(charCF)
            } else {
                dispAlphabet.indexOf(charCF)
            }
        }
        return alphabet[result % 26].toString()
    }

    private fun calcolaCognome(cognome: String): String {
        val uppedCognome = cognome.toUpperCase()
        var result = String()
        if (uppedCognome.length < 3) {
            result = uppedCognome
            for (i in uppedCognome.length..2) {
                result += "X"
            }
            return result
        }
        var i = 0
        while (i < uppedCognome.length && result.length < 3) {
            if (isConsonante(uppedCognome[i])) {
                result += uppedCognome[i]
            }
            i++
        }
        if (result.length < 3) {
            var i = 0
            while (i < uppedCognome.length && result.length < 3) {
                if (isVocale(uppedCognome[i])) {
                    result += uppedCognome[i]
                }
                i++
            }
        }
        return result
    }

    private fun calcolaNome(nome: String): String {
        val uppedNome = nome.toUpperCase()
        var result = String()
        if (uppedNome.length < 3) {
            result = uppedNome
            for (i in uppedNome.length..2) {
                result += "X"
            }
            return result
        }
        var i = 0
        while (i < uppedNome.length && result.length < 4) {
            if (isConsonante(uppedNome[i])) {
                result += uppedNome[i]
            }
            i++
        }
        return if (result.length < 4) {
            calcolaCognome(uppedNome)
        } else result
    }

    private fun isVocale(c: Char): Boolean {
        return vocali.contains(c.toString())
    }

    private fun isConsonante(c: Char): Boolean {
        return !isVocale(c)
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val CodiceFiscale = CodiceFiscale()
            println(CodiceFiscale.calcolaCodiceFiscale(
                    "Mario", "Rossi", 12, 6, 1976, "M", "Bologna") == "RSSMRA76H12A944I")
            println(CodiceFiscale.calcolaCodiceFiscale(
                    "Mario", "Rossi", 1, 1, 1990, "M", "Milano") == "RSSMRA90A01F205Z")
            println(CodiceFiscale.verificaCodiceFiscale(
                    "Mario", "Rossi", 12, 6, 1976, "M", "Bologna", "RSSMRA76H12A94QF"))
            println(CodiceFiscale.verificaCodiceFiscale(
                    "Mario", "Rossi", 12, 6, 1976, "M", "Bologna", "RSSMRA76H12A9Q4U"))
            println(CodiceFiscale.verificaCodiceFiscale(
                    "Mario", "Rossi", 1, 1, 1990, "M", "Milano", "RSSMRA90A01F20RU"))
            println(CodiceFiscale.verificaCodiceFiscale(
                    "Mario", "Rossi", 1, 1, 1990, "M", "Milano", "RSSMRA90A01F2L5K"))
            println(CodiceFiscale.verificaCodiceFiscale(
                    "Mario", "Rossi", 1, 1, 1990, "M", "Milano", "RSSMRA90A01FN05O"))
            println(CodiceFiscale.verificaCodiceFiscale(
                    "Mario", "Rossi", 1, 1, 1990, "M", "Milano", "RSSMRA90A0MF205R"))
            println(CodiceFiscale.verificaCodiceFiscale(
                    "Mario", "Rossi", 1, 1, 1990, "M", "Milano", "RSSMRA90AL1F205K"))
        }
    }
}